server/src/main/java/io/crate/expression/scalar/timestamp/TimeFunction.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to CRATE Technology GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.expression.scalar.timestamp;
+
+import io.crate.data.Input;
+import io.crate.metadata.FunctionIdent;
+import io.crate.metadata.FunctionInfo;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataTypes;
+
+import javax.annotation.Nullable;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.BiFunction;
+
+abstract class TimeFunction<R> extends Scalar<R, Integer> {
+
+    private final Signature signature;
+    private final FunctionInfo info;
+    private final int defaultPrecision;
+    private final BiFunction<Instant, Integer, R> logic;
+
+    TimeFunction(Signature signature, int defaultPrecision, BiFunction<Instant, Integer, R> logic) {
```

Comment:
It's kinda hard to understand what `Integer, Integer` here are referring to in `logic` 

Why the big diverge from the prior solution?

Reply:
There are three related functions:

Reply:
I feel like this is confusing boilerplate (creating signature *data* structures) with duplicating logic.

Yes, `NowFunction` and `CurrentTimestampFunction` could probably share the same implementation for the zero argument case - but in that case they could actually return the exact same instance, instead of introducing a common base class.


`CurrentTimeFunction` has little to do with `current_timestamp`. The signatures look similar - but that's about it.

The `TimeFunction` instances are not interchangeable (substitution principle). So the reason for this abstraction is only to "share common code" which is the signature definitions.
There are better ways than inheritance to achieve that.